### PR TITLE
esp8266: Build the SDK bootloader from source.

### LINKED ISSUE #16402: esp8266 precompiled bootloaders don't support partitions past 1MB
#### Description
The esp8266 cpu ships with some bootloader code pre-compiled from the vendor SDK. I couldn't find what revision and options where used to compile these bootloaders, they have the following bug:

If the partition table contains any partition that starts or ends at the 1MB boundary or after that you get the following error even if using a chip with 4MB flash:

```
I (69) boot: SPI Flash Size : 4MB
E (72) flash_parts: partition 3 invalid - offset 0x100000 size 0x2000 exceeds flash chip size 0x100000
E (81) boot: Failed to verify partition table
E (85) boot: load partition table error!
user code done
```

Note that the code detects the flash at 4MB but then complains about the partition being larger than 1MB.

#### Steps to reproduce the issue
Create a custom partition table with the vendor recommended OTA scheme, for example:

```
$ cat mypartitions.csv
# Name,   Type, SubType, Offset,   Size
nvs,      data, nvs,     0x9000,   0x6000
phy_init, data, phy,     0xf000,   0x1000
ota_0,    app,  ota_0,   0x010000, 0xf0000
otadata,  data, ota,     0x100000, 0x2000
ota_1,    app,  ota_1,   0x110000, 0xf0000
```

Then compile and run on a esp12 board the following example program:

```USEMODULE=esp_log_startup make -C tests/shell BOARD=esp8266-esp-12x PARTITION_TABLE_CSV=`pwd`/mypartitions.csv -j flash```

#### Expected results
The board should boot all the way to the shell.

#### Actual results
The ROM bootloader loads the flash bootloader at 0x0000 and prints messages in the UART at 74880 bps while doing that; and then the flash bootloader switches to 115200 and prints the following error:

```
I (69) boot: SPI Flash Size : 4MB
E (72) flash_parts: partition 3 invalid - offset 0x100000 size 0x2000 exceeds flash chip size 0x100000
E (81) boot: Failed to verify partition table
E (85) boot: load partition table error!
```

### PR DESCRIPTION
**Pull request by iosabi**
### Contribution description

We had four versions of pre-built bootloaders for the esp8266 with different settings of logging and color logging. These bootloaders were manually built from the SDK and shipped with RIOT-OS source code. However there are more settings that affect the bootloader build that are relevant to the app or final board that uses this bootloader. In particular, flash size and flash speed is important for the bootloader to be able to load an app from a large partition table at the fastest speed supported by the board layout and flash chip.

Another example is the UART baudrate of the logging output from the bootloader. The boot ROM will normally start at a baud rate of 74880 (depending on the crystal installed), so it might make sense to keep the UART output at the same speed so we can debug boot modes and bootloader with the same terminal.

This patch builds the `bootloader.bin` file from the ESP8266 SDK source code. The code is built as a module (`esp8266_bootloader`) which at the moment doesn't generate any object code for the application and only produces a `bootloader.bin` file set to the `BOOTLOADER_BIN` make variable for the `esptool.inc.mk` to flash.

The code needs to be compiled and linked with custom rules defined in the module's Makefile since the `bootloader.bin` is its own separate application.

The `BOOTLOADER_BIN` variable is changed from a path relative to the `$(RIOTCPU)/$(CPU)/bin/` directory to be full path. This makes it easier for applications or board to provide their own bootloader binary if needed.

As a result of building the bootloader from source we fixed the issue of having a large partition table.
### Testing procedure
This on a esp12x shows the ROM bootloader log output (`ets Jan  8 2013,rst cause:2, boot mode:(3,6)`) and the sencond-stage bootloader built by this PR (`ESP-IDF v3.1-51-g913a06a9`) at 74880 baudrate. The application output is seen as garbage since the esp8266-esp-12x uses 115200 as baurate by default.
```
USEMODULE=esp_log_startup make -C tests/shell BOARD=esp8266-esp-12x flash term BAUD=74880 RIOT_TERMINAL=picocom
```

The same command but with `BAUD=115200` (or the BAUD parameter missing) shows the second-stage bootloader output and then the application output at 115200.

### Issues/PRs references
Fixes #16402.

### CODE REVIEW

File: pkg/esp8266_sdk/bootloader/sdkconfig.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2021 iosabi
```
[the remark this answers is not in the document]
Comment (iosabi) [reply pkg/esp8266_sdk/bootloader/sdkconfig.h:2]: It doesn't make any difference for the purpose of copyright protection. Also having individuals listed at the top of each file is not very accurate since the copyright applies to the specific contribution: if I add a function to somebody else's copyrighted file I hold the copyright for that function and they hold the copyright for the rest. When suggestions are made in code reviews in the form of patches that's even more debatable. Many projects just list copyright "the [project] authors" to avoid all this anyway since it doesn't make any difference. I'd prefer to leave it like this.

**Comment by iosabi** [comment]
Renamed `esp8266_bootloader` to `esp_bootloader`. For the SDK, I expect to have `esp_sdk` or `esp_idf_sdk` for the esp-idf repo instead of the old esp8266 one, eventually.

File: pkg/esp8266_sdk/bootloader/sdkconfig.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig.h:2]: Unfortunately, I don't know what the policy of the project owners is. At least there is a statement in the [coding conventions](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md) there is a statement:
```
* All files contain the copyright note and the author.
```
whatever author means in this context. I haven't seen authors use just a synonym or their GitHub user yet. Maybe @kaspar030 as a project owner can clarify.

**Comment by gschorcht** [comment]
> I expect to have `esp_sdk` or `esp_idf_sdk`

In my opinion, both would be fine.

File: pkg/esp8266_sdk/bootloader/sdkconfig.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig.h:2]: @kaspar030 ping

[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig.h:2]: Maybe @tcschmidt as a project owner can clarify the policy on using the author's name in copyright notices for this project? I would like to move forward with this PR.

**Comment by gschorcht** [comment]
Please squash.

**Comment by benpicco** [comment]
looks like CI is not happy

**Comment by gschorcht** [comment]
Hm, all compilations fail with error:
```
make: Entering directory 'tests/driver_l3g4200d'
make: *** No rule to make target 'build/esp_bootloader/bootloader.bin', needed by 'test-input-hash'.  Stop.
make: *** Waiting for unfinished jobs....
Building application "tests_driver_l3g4200d" for "esp8266-sparkfun-thing" with MCU "esp8266".
rm -rf build/pkg-build/esp8266_sdk

make: Leaving directory 'tests/driver_l3g4200d'
cat: build/test-input-hash.sha1: No such file or directory
```
Local compilations work without any problems. The SDK is pulled and the bootloader is built and available as expected for example `tests/driver_l3g4200d/bin/esp8266-sparkfun-thing/esp_bootloader/bootloader.bin`. I have noidea why the compilation in CI is expecting to have `build/esp_bootloader/bootloader.bin`.

**Comment by benpicco** [comment]
This appears to be caused by [this check](https://github.com/RIOT-OS/RIOT/blob/master/makefiles/tests/tests.inc.mk#L74).
Not sure how the bootloader ends up in there

File: pkg/esp8266_sdk/Makefile.include [hunk 2]
```diff
@@ -21,3 +21,13 @@ LINKFLAGS += -L$(ESP8266_SDK_BUILD_DIR)
...
 PSEUDOMODULES += esp8266_sdk
+
+# Bootloader module built from the SDK.
+DIRS += $(RIOTBASE)/pkg/esp8266_sdk/bootloader
+PSEUDOMODULES += esp_bootloader
+
+ifneq (,$(filter esp_bootloader,$(USEMODULE)))
+  # Bootloader file used by esptool.inc.mk
+  BOOTLOADER_BIN ?= $(BINDIR)/esp_bootloader/bootloader.bin
+  TEST_EXTRA_FILES += $(BOOTLOADER_BIN)
```
Comment (fjmolinas) [inline pkg/esp8266_sdk/Makefile.include:32]: There is a name clash between this an `BOOTLOADER_BIN` defined in `riotboot.mk`, can we rename it to `ESP_BOOTLOADER_BIN` (I'm not sure that is the issue though)

Also wouldn't `BOOTLOADER_BIN` need to be added to `BUILD_FILES`?

Comment (fjmolinas) [reply pkg/esp8266_sdk/Makefile.include:32]: The error message is also correct... what is the rule to generate `BOOTLOADER_BIN`?

```
make: Entering directory '/tmp/dwq.0.037298659752877716/749aff2d092a21a8cc4cbf9b746da3f6/examples/timer_periodic_wakeup'
make: *** No rule to make target '/tmp/dwq.0.037298659752877716/749aff2d092a21a8cc4cbf9b746da3f6/build/esp_bootloader/bootloader.bin', needed by 'test-input-hash'.  Stop
```

Comment (fjmolinas) [reply pkg/esp8266_sdk/Makefile.include:32]: Should the actual file to be added be `ESP_SDK_ESP_BOOTLOADER_BIN`? The whole dependency is a bit hard for me to grasp to suggest a fix for the issues on the CI...

File: pkg/esp8266_sdk/bootloader/sdkconfig_default.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Generated by ./update_mk.sh, don't modify directly.
```
Comment (gschorcht) [inline pkg/esp8266_sdk/bootloader/sdkconfig_default.h:2]: @iosabi Is this file really needed as part of the repository or could the initial version also be generated if not existing?

File: pkg/esp8266_sdk/Makefile.include (hunk 2 above)
Comment (gschorcht) [reply pkg/esp8266_sdk/Makefile.include:32]: > There is a name clash between this an `BOOTLOADER_BIN` defined in `riotboot.mk`, can we rename it to `ESP_BOOTLOADER_BIN` (I'm not sure that is the issue though)

Hm, the variable for the bootloader binary was already called `BOOTLOADER_BIN` before this PR. Thus, I wouldn't suppose that this causes the problem. But we should think about to change it to something more specific.

> Also wouldn't `BOOTLOADER_BIN` need to be added to `BUILD_FILES`?

`BUILD_FILES` is the variable which contains all binary files that are required to flash a module without having installed the toolchain, right? 

> Should the actual file to be added be `ESP_SDK_ESP_BOOTLOADER_BIN`?

I don't think so because this variable is just used during the compilation of the bootloader.

Comment (fjmolinas) [reply pkg/esp8266_sdk/Makefile.include:32]: > Hm, the variable for the bootloader binary was already called BOOTLOADER_BIN before this PR. Thus, I wouldn't suppose that this causes the problem. But we should think about to change it to something more specific.

Yes but now its added to `TEST_EXTRA_FILES` which was not the case before, and its this addition that for some reason triggers the failed check.

File: pkg/esp8266_sdk/bootloader/sdkconfig_default.h (hunk 3 above)
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig_default.h:2]: @iosabi ping

Comment (iosabi) [reply pkg/esp8266_sdk/bootloader/sdkconfig_default.h:2]: I think I wrote a comment somewhere else about this. I looked into it but the issue is that to generate this file you need to have both the host toolchain (to run the "make defconfig") and the target esp8266 toolchain. I don't think that any of these settings are relevant, they are just used by the sources compiled and must be defined but the defaults are ok. The interesting ones are in `pkg/esp8266_sdk/bootloader/sdkconfig.h`, and if you add more there (tied to some RIOT config) the script will remove them from the `sdkconfig_default.h`. I think generating it when building RIOT was too technically complex to warrant it since a static file is enough. For the .mk file with the list of sources I'm not sure it is possible or easy.

File: pkg/esp8266_sdk/bootloader/sdkconfig.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig.h:2]: No further concerns.

File: pkg/esp8266_sdk/bootloader/sdkconfig_default.h (hunk 3 above)
Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig_default.h:2]: @iosabi Please consider to merge and push my changes to your branch. We can try again to make Murdock happy.

Comment (gschorcht) [reply pkg/esp8266_sdk/bootloader/sdkconfig_default.h:2]: @iosabi With PR #17457, the concept of compiling the boot loader from package sources works, including adding the binary to `TEST_EXTRA_FILES`. Murdock is satisfied. So the changes I provided in https://github.com/iosabi/RIOT/pull/1 should work as well.
If it works for both platforms, we could move some stuff to `cpu/esp_common`.

**Comment by gschorcht** [comment]
Ping @iosabi I would like to get this PR in the next release. Could you merge my changes into your branch, please. The same works for ESP32, PR #17457.

**Comment by gschorcht** [comment]
Ping @iosabi The code freeze for next release is over but I'm still interested to get it merged. If you don't want to continue this work, would it be ok if I take it over?

**Comment by benpicco** [comment]
I think it's fine to push your patch to iosabi/esp8266_bldr, have me ACK it and then squash - no need to let this wait :wink:

**Comment by iosabi** [comment]
Sorry I'm not able to follow up on this right now. Feel free to take over the branch if you need to push something there.

**Comment by gschorcht** [comment]
I will take over this PR once I have time for it.

**Comment by gschorcht** [comment]
Reopened as PR #19073 rebased onto current master with some fixes.
